lib/charms/alertmanager_k8s/v0/alertmanager.py

```diff
@@ -0,0 +1,132 @@
+        """This hook stores locally the address of the newly-joined alertmanager.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        if event.unit:  # event.unit may be `None` in the case of app data change
+            # Save locally the public IP address of the alertmanager unit
+            if address := event.relation.data[event.unit].get("public_address"):
+                # TODO consider storing in unit data instead of StoredState
+                self._stored.alertmanagers[event.unit.name] = address
+
+                # inform consumer about the change
+                # manually emitting the "available" event until scale-up/scale-down "ready" flag is fixed
+                self.on.available.emit()
+
+    def get_cluster_info(self) -> List[str]:
+        """Returns a list of ip addresses of all the alertmanager units
+        """
+        return sorted(list(self._stored.alertmanagers.values()))
+
+    def _on_relation_departed(self, event: ops.charm.RelationDepartedEvent):
+        """This hook removes the address of the departing alertmanager from its local store.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        if self._stored.alertmanagers.pop(event.unit.name, None):
+            # inform consumer about the change
+            # manually emitting the "available" event until scale-up/scale-down "ready" flag is fixed
+            self.on.available.emit()
+
+    def _on_relation_broken(self, event: ops.charm.RelationBrokenEvent):
+        self._stored.alertmanagers.clear()
+        # inform consumer about the change
+        # manually emitting the "available" event until scale-up/scale-down "ready" flag is fixed
+        self.on.available.emit()
+
+
+class AlertmanagerProvider(ProviderBase):
+    """A "provider" handler to be used by the Alertmanager charm for abstracting away all the
+    communication with consumers.
+    This provider auto-registers relation events on behalf of the main Alertmanager charm.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            service_name (str): a name for the provided service
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): the Alertmanager charm
+    """
+    _provider_relation_name = "alerting"
+    _public_api_port: int = 9093  # public port may not be the same as AlertmanagerCharm._api_port
+
+    def __init__(self, charm, service_name: str, version: str = None):
+        super().__init__(charm, self._provider_relation_name, service_name, version)
+        self.charm = charm
+        self._service_name = service_name
+
+        events = self.charm.on[self._provider_relation_name]
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
```

Comment:
Isn't like 116 a duplicate of line 120 or am I missing something ?

Reply:
Removed.